Add option to use ssl on stale checks

jvandemark opened the pull request:
Added option (--ssl) to use SSL communication when running check-stale-results.rb

## Pull Request Checklist

**Is this in reference to an existing issue?**
no

#### General

- [x] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [x] Update README with any necessary configuration snippets

- [ ] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [ ] Tests

- [ ] Add the plugin to the README

- [ ] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

#### Purpose
Add option to use ssl on stale checks

#### Known Compatibility Issues

jvandemark committed:
Add option to use ssl on stale checks

Added option (--ssl) to use SSL communication when running check-stale-results.rb

majormoses commented:
While it is nice to add it as an explicit option I am pretty sure you can do this by just setting `settings['api']['host']` to include `https` as a protocol.

majormoses commented:
At some point I am gonna run through all the plugins and rename any options such as `ssl`, `scheme`, etc that are really all just a protocol and standardize them. I'd prefer if we renamed it now rather than later. This is generally the prefered way of defining it: https://github.com/sensu-plugins/sensu-plugins-graylog/blob/1.3.1/bin/check-graylog-streams.rb#L29-L33

jvandemark commented:
Thanks for the feedback, I will try using settings['api']['host']. If it works then this pull request can be cancelled. 

----- 

Unfortunately this approach doesn't appear to work. It would work if the code were using a uri.

uri = URI('https://secure.example.com')

Net::HTTP.start(uri.host, uri.port)

uri.host returns only the host 'secure.example.com'

jvandemark commented:
Just so I understand you would prefer the option 'protocol' be used?

I will make the change and push again, thanks

jvandemark committed:
change option to protocol

jvandemark commented:
these change were manually tested with some additional debug, which was removed for the submission

debug output: https://gist.github.com/jvandemark/5208b14f395237c85a038c51b701332d

majormoses commented:
Another approach from the ruby docs is to inspect the uri scheme like this:
```ruby
uri = URI('https://secure.example.com/some_path?query=string')

Net::HTTP.start(uri.host, uri.port,
  :use_ssl => uri.scheme == 'https') do |http|
  request = Net::HTTP::Get.new uri

  response = http.request request # Net::HTTPResponse object
end
```
https://docs.ruby-lang.org/en/2.0.0/Net/HTTP.html#class-Net::HTTP-label-HTTPS

One of the benefits here is that it just works without any additional args on the other hand it means that a user needs to know to put https in their URL. I am not convinced one is better than the other but I wanted to see what the ruby docs said because I felt like it should have just worked passing the protocol/scheme.

I prefer using protocol over scheme in terminology for monitoring because it causes confusion with metrics scheme which is why I prefer protocol as it is clear to it's intent without the ambiguity.

jvandemark committed:
Refactor to use URI based on plugin configuration

Remove protocol option
Allow protocol to be defined as part of the plugin configuation
If the protocol is not part of the configuration default to http
Construct URI based on plugin configuration
Remove method argument from api_request as it's no longer needed

jvandemark commented:
I refactored the submission to use a URI, and moved the ability to set the protocol to the plugin configuration.

majormoses commented on bin/check-stale-results.rb:
```diff
@@ -60,22 +61,27 @@ def humanize(secs)
     end.compact.reverse.join(' ')
   end
 
-  def api_request(method, path)
```
why remove the method? I get that this probably will never change but I see no value in removing it.

jvandemark replied:
Fair enough :) I changed it for the reason you noted. I will add it back.

jvandemark committed:
add method to manage protocol and uri construction

jvandemark committed:
refactoring as per code review

remove un-needed local variable declaration
refactored the method get_uri to use https if it's part of the host

jvandemark committed:
refactoring as per code review

remove un-needed local variable declaration
refactored the method get_uri to use https if it's part of the host
update CHANGELOG.md

jvandemark committed:
Code changes as per code review

jvandemark committed:
Update check-stale-results.rb

majormoses commented on bin/check-stale-results.rb:
```diff
@@ -60,22 +61,32 @@ def humanize(secs)
-  def api_request(method, path)
+  def get_uri(path)
+    protocol = (settings['api'].key?('protocol') ? settings['api']['protocol'] : 'http')
+    uri = URI(protocol + '://' + settings['api']['host'] + ':' + settings['api']['port'].to_s + path)
```
It's better to use string interpolation over string concatenation for the following reasons:
- string concatenation requires creating 3 strings for combining 2 strings therefore is more exspensive
- it automatically calls `.to_s` for you so there is no need to do that
- it is more common practice in the community
- more idiomatic
- less typing
- better use of chars on a line
```ruby
URI("#{protocol}://#{settings['api']['host']}:#{settings['api']['port']}#{path}")
```

Also you don't need to assign it to a var if you are just returning it you you can remove the `url` below and remove the assignment.

jvandemark replied:
Nice, thanks, changes made.
I've spent most of my career writing java so Ruby is "new" and I don't know it that well.

majormoses replied:
I view open source as one big playground of learning. This is why code review happens, we all learn from things and improve the code. I spent most of my career with bash and python up until a few years ago, so I definitely understand that you often bring familiar patterns with you. I use to write really great python in my ruby, now I write great ruby in my python. Each language has its own little quirks and contributing to open source is a great way to learn them as someone will generally point out something you did no know.

majormoses commented on bin/check-stale-results.rb:
```diff
@@ -60,22 +61,32 @@ def humanize(secs)
   def results
     res = []
-    req = api_request(:GET, '/results')
+    req = api_request('/results')
```
we remove the parameterization?

jvandemark replied:
reverted

majormoses commented on bin/check-stale-results.rb:
```diff
@@ -60,22 +61,32 @@ def humanize(secs)
+    uri
+  end
+
+  def api_request(path)
```
why remote the flexibility of specifying the request type?

jvandemark replied:
reverted

jvandemark commented:
All requested changes made, thanks for the feedback

majormoses commented on bin/check-stale-results.rb:
```diff
@@ -60,17 +61,26 @@ def humanize(secs)
-    if settings['api']['user'] && settings['api']['password']
-      req.basic_auth(settings['api']['user'], settings['api']['password'])
+    uri = get_uri(path)
+    response = nil
```
what does this serve? We assign it further down a little bit further to something else. In ruby you can assign vars on the fly no need to set them up beforehand.

jvandemark replied:
Ah, makes sense. Changed as per your suggestion.

majormoses commented on bin/check-stale-results.rb:
```diff
@@ -60,17 +61,26 @@ def humanize(secs)
   end
 
+  def get_uri(path)
+    protocol = (settings['api'].key?('protocol') ? settings['api']['protocol'] : 'http')
```
I am thinking that as `settings['api]['protocol']` is not standard we should either inspect the host to see if the protocol is embedded or expose it as a cli arg. for the host, user, password it makes sense to re-use those as they are pretty much guaranteed to be there. Matter of preference really but just my 2 cents.

jvandemark replied:
I refactored the method `get_uri` to extract the protocol (https only) if it's part of the host.

majormoses reviewed (commented):
Couple comments, let me know if you need any more help.

jvandemark commented:
Changed as per your suggestions

majormoses reviewed (approved):
LGTM

majormoses commented:
released: https://rubygems.org/gems/sensu-plugins-sensu/versions/2.5.0
